### PR TITLE
Listview reordering updatelayout

### LINKED ISSUE #16799: ListView Drag and Drop Items - Error when CompleteReordering
### Current behavior

When we created some ListView using some list of Records.
And tryed to Reorder the Items on the ListView we are getting some errors

```
fail: Microsoft.UI.Xaml.UIElement[0]
      Failed to raise 'DropEvent': System.NullReferenceException: Object reference not set to an instance of an object.
         at Microsoft.UI.Xaml.Controls.VirtualizingPanelLayout.RecycleLine(Line line, Boolean clearContainer) in /__w/1/s/src/Uno.UI/UI/Xaml/Controls/ListViewBase/VirtualizingPanelLayout.managed.cs:line 558
         at Microsoft.UI.Xaml.Controls.VirtualizingPanelLayout.ClearLines(Boolean clearContainer) in /__w/1/s/src/Uno.UI/UI/Xaml/Controls/ListViewBase/VirtualizingPanelLayout.managed.cs:line 835
         at Microsoft.UI.Xaml.Controls.VirtualizingPanelLayout.Refresh() in /__w/1/s/src/Uno.UI/UI/Xaml/Controls/ListViewBase/VirtualizingPanelLayout.managed.cs:line 818
         at Microsoft.UI.Xaml.Controls.VirtualizingPanelLayout.CompleteReorderingItem(FrameworkElement element, Object item) in /__w/1/s/src/Uno.UI/UI/Xaml/Controls/ListViewBase/VirtualizingPanelLayout.managed.cs:line 1158
         at Microsoft.UI.Xaml.Controls.ListViewBase.CompleteReordering(FrameworkElement draggedContainer, Object draggedItem) in /__w/1/s/src/Uno.UI/UI/Xaml/Controls/ListViewBase/ListViewBase.DragDrop.cs:line 392
         at Microsoft.UI.Xaml.Controls.ListViewBase.OnReorderCompleted(Object sender, DragEventArgs dragEventArgs) in /__w/1/s/src/Uno.UI/UI/Xaml/Controls/ListViewBase/ListViewBase.DragDrop.cs:line 261
         at Microsoft.UI.Xaml.UIElement.InvokeHandler(Object handler, RoutedEventArgs args) in /__w/1/s/src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs:line 982
         at Microsoft.UI.Xaml.UIElement.InvokeHandlers(IList`1 handlers, RoutedEventArgs args, Boolean& isHandled) in /__w/1/s/src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs:line 933
         at Microsoft.UI.Xaml.UIElement.RaiseEvent(RoutedEvent routedEvent, RoutedEventArgs args, BubblingContext ctx) in /__w/1/s/src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs:line 688
         at Microsoft.UI.Xaml.UIElement.RaiseOnParent(RoutedEvent routedEvent, RoutedEventArgs args, UIElement parent, BubblingContext ctx) in /__w/1/s/src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs:line 757
         at Microsoft.UI.Xaml.UIElement.RaiseEvent(RoutedEvent routedEvent, RoutedEventArgs args, BubblingContext ctx) in /__w/1/s/src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs:line 688
         at Microsoft.UI.Xaml.UIElement.RaiseOnParent(RoutedEvent routedEvent, RoutedEventArgs args, UIElement parent, BubblingContext ctx) in /__w/1/s/src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs:line 757
         at Microsoft.UI.Xaml.UIElement.RaiseEvent(RoutedEvent routedEvent, RoutedEventArgs args, BubblingContext ctx) in /__w/1/s/src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs:line 688
         at Microsoft.UI.Xaml.UIElement.RaiseOnParent(RoutedEvent routedEvent, RoutedEventArgs args, UIElement parent, BubblingContext ctx) in /__w/1/s/src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs:line 757
         at Microsoft.UI.Xaml.UIElement.RaiseEvent(RoutedEvent routedEvent, RoutedEventArgs args, BubblingContext ctx) in /__w/1/s/src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs:line 688
         at Microsoft.UI.Xaml.UIElement.RaiseOnParent(RoutedEvent routedEvent, RoutedEventArgs args, UIElement parent, BubblingContext ctx) in /__w/1/s/src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs:line 757
         at Microsoft.UI.Xaml.UIElement.RaiseEvent(RoutedEvent routedEvent, RoutedEventArgs args, BubblingContext ctx) in /__w/1/s/src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs:line 688
         at Microsoft.UI.Xaml.UIElement.RaiseOnParent(RoutedEvent routedEvent, RoutedEventArgs args, UIElement parent, BubblingContext ctx) in /__w/1/s/src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs:line 757
         at Microsoft.UI.Xaml.UIElement.RaiseEvent(RoutedEvent routedEvent, RoutedEventArgs args, BubblingContext ctx) in /__w/1/s/src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs:line 688
         at Microsoft.UI.Xaml.UIElement.SafeRaiseEvent(RoutedEvent routedEvent, RoutedEventArgs args, BubblingContext ctx) in /__w/1/s/src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs:line 580
```

Or in Debug.

image

Works fine on WinAppSDK but not on Desktop platform.

When Start.

image

When Drag.

image

When Drop
image

### Expected behavior

The expectation is that the list is reordered correctly.
We need to use the event that observes the order of the elements in the list.

### How to reproduce it (as minimally and precisely as possible)

There is a project with the scenario in which the error happens.

[UnoAppListViewDrag.zip](https://github.com/unoplatform/uno/files/15380761/UnoAppListViewDrag.zip)

Using UnoApp.
Creating a ListView on the Main page.
Including a list that has record as one of its attributes.

### Workaround

_No response_

### Works on UWP/WinUI

Yes

### Environment

_No response_

### NuGet package version(s)

  <PropertyGroup>
    <UnoExtensionsVersion>4.2.0-dev.77</UnoExtensionsVersion>
    <UnoToolkitVersion>6.1.0-dev.24</UnoToolkitVersion>
    <UnoThemesVersion>5.1.0-dev.26</UnoThemesVersion>
    <UnoCSharpMarkupVersion>5.3.0-dev.70</UnoCSharpMarkupVersion>
  </PropertyGroup>

### Affected platforms

_No response_

### IDE

Visual Studio 2022

### IDE version

Microsoft Visual Studio Community 2022 (2) (64-bit) - Current Version 17.9.0

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by ramezgerges**
GitHub Issue (If applicable): closes #16799

## PR Type

What kind of change does this PR introduce?

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/Controls/ListViewBase/VirtualizingPanelLayout.managed.cs [hunk 1]
```diff
@@ -830,11 +835,24 @@ private void ClearLines(bool clearContainer)
...
+				// Recycling a line can lead to removing elements from the visual tree.
+				// If a user forces a layout during this (e.g. in Unloaded), we will
+				// update the _materializedLines in MeasureOverride. So, to prevent
+				// modification while looping, we take a copy and clear before looping.
+				var lines = _materializedLines.ToList();
+				_materializedLines.Clear();
+				foreach (var line in lines)
+				{
+					RecycleLine(line, clearContainer);
+				}
```
Comment (Xiaoy312) [inline src/Uno.UI/UI/Xaml/Controls/ListViewBase/VirtualizingPanelLayout.managed.cs:850]:
```suggestion
				// Recycling a line can lead to elements removal from the visual tree.
				// To prevent modification while looping, we take a copy of _materializedLines
				// and clear the original before starting the loop. This ensures that if a user forces 
				// a layout during this process (e.g., in Unloaded), we can update _materializedLines 
				// in MeasureOverride without issues.
				var lines = _materializedLines.ToArray();
				_materializedLines.Clear();
				foreach (var line in lines)
				{
					RecycleLine(line, clearContainer);
				}
```

File: src/Uno.UI/UI/Xaml/Controls/ListViewBase/VirtualizingPanelLayout.managed.cs [hunk 2]
```diff
@@ -65,11 +65,12 @@ private protected VirtualizingPanelGenerator Generator
 		/// <summary>
 		/// Ordered record of all currently-materialized lines.
 		/// </summary>
-		private readonly Deque<Line> _materializedLines = new Deque<Line>();
+		private Deque<Line> _materializedLines { get; } = new Deque<Line>();
```
Comment (Xiaoy312) [inline src/Uno.UI/UI/Xaml/Controls/ListViewBase/VirtualizingPanelLayout.managed.cs:68]: unnecessary change
```suggestion
		private readonly Deque<Line> _materializedLines = new Deque<Line>();
```

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ListViewBase.cs [hunk 3]
```diff
@@ -4591,6 +4591,59 @@ public async Task When_ScrollIntoView_Containers_With_Varying_Heights()
...
+			using var mouse = injector.GetMouse();
+
+			mouse.MoveTo(SUT.GetAbsoluteBoundsRect().GetCenter() with { Y = SUT.GetAbsoluteBoundsRect().Y + 50 });
+			mouse.Press();
+			await WindowHelper.WaitForIdle();
+
+			mouse.MoveTo(SUT.GetAbsoluteBoundsRect().GetCenter() with { Y = SUT.GetAbsoluteBoundsRect().Y + 150 });
+			await WindowHelper.WaitForIdle();
+			mouse.Release();
+			await WindowHelper.WaitForIdle();
```
Comment (Xiaoy312) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ListViewBase.cs:4636]:
```suggestion
			// drag(pick-up) item#0
			mouse.MoveTo(SUT.GetAbsoluteBoundsRect().GetCenter() with { Y = SUT.GetAbsoluteBoundsRect().Y + 50 });
			await WindowHelper.WaitForIdle();
			mouse.Press();
			await WindowHelper.WaitForIdle();
			
			// drop onto item#1
			mouse.MoveTo(SUT.GetAbsoluteBoundsRect().GetCenter() with { Y = SUT.GetAbsoluteBoundsRect().Y + 150 });
			await WindowHelper.WaitForIdle();
			mouse.Release();
			await WindowHelper.WaitForIdle();
```

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ListViewBase.cs [hunk 4]
```diff
@@ -4591,6 +4591,59 @@ public async Task When_ScrollIntoView_Containers_With_Varying_Heights()
...
+			{
+				SUT.Items.Add(new UpdateLayoutOnUnloadedControl
+				{
+					Content = new TextBlock
+					{
+						AllowDrop = true,
+						CanDrag = true,
+						Margin = new Thickness(50),
+						Text = i.ToString()
+					}
```
Comment (Xiaoy312) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ListViewBase.cs:4620]: would be better to explicitly set the height, than margin
```suggestion
					Content = new TextBlock
					{
						AllowDrop = true,
						CanDrag = true,
						Height = 100,
						Text = i.ToString()
					}
```

**Review by Xiaoy312 — APPROVED** [review]

**Review by MartinZikmund — APPROVED** [review]

**Review by jeromelaban — APPROVED** [review]
